File: frontend/src/types/itinerary.ts
export interface Place {
  id: string
  google_place_id: string
  name: string
  lat: number
  lng: number
  types: string[]
  opening_hours?: OpeningHours
  price_level: number
  rating: number
  address: string
}

export interface OpeningHours {
  periods: Period[]
}

export interface Period {
  day_of_week: number
  open_time: string
  close_time: string
}

export interface Itinerary {
  id: string
  title: string
  destination: string
  start_date: string
  end_date: string
  days: DayPlan[]
  created_at: string
  updated_at: string
}

export interface DayPlan {
  day_number: number
  date: string
  activities: Activity[]
}

export interface Activity {
  order: number
  place_id: string
  place?: Place
  start_time: string
  end_time: string
  duration_min: number
  travel_from_prev?: TravelSegment
  note?: string
  estimated_cost_yen: number
}

export interface TravelSegment {
  mode: 'walk' | 'train' | 'bus' | 'taxi' | 'driving'
  duration_min: number
  distance_m: number
  estimated_cost_yen: number
}

export interface Violation {
  type: string
  day_number: number
  activity_idx: number
  message: string
}

export interface PlanRequest {
  destination: string
  num_days: number
  start_date: string
  preferences: {
    interests: string[]
    budget: 'budget' | 'moderate' | 'luxury'
    travel_style: 'relaxed' | 'active' | 'balanced'
    total_budget_yen?: number
  }
  constraint: {
    max_walk_distance_m: number
    max_activities_day: number
    earliest_start_time: string
    latest_end_time: string
  }
}

export interface DayCost {
  day_number: number
  cost_yen: number
}

export interface BudgetSummary {
  total_cost_yen: number
  daily_costs: DayCost[]
}

export interface GenerateResult {
  itinerary: Itinerary
  violations: Violation[]
  budget_summary: BudgetSummary
}

File: api/internal/domain/itinerary.go
package domain

import "time"

// Itinerary represents a complete travel plan.
type Itinerary struct {
	ID          string       `json:"id"`
	Title       string       `json:"title"`
	Destination string       `json:"destination"`
	StartDate   time.Time    `json:"start_date"`
	EndDate     time.Time    `json:"end_date"`
	Days        []DayPlan    `json:"days"`
	CreatedAt   time.Time    `json:"created_at"`
	UpdatedAt   time.Time    `json:"updated_at"`
}

// DayPlan represents a single day's plan within an itinerary.
type DayPlan struct {
	DayNumber  int        `json:"day_number"`
	Date       time.Time  `json:"date"`
	Activities []Activity `json:"activities"`
}

// Activity represents a single activity in a day plan.
type Activity struct {
	Order         int    `json:"order"`
	PlaceID       string `json:"place_id"`
	Place         *Place `json:"place,omitempty"`
	StartTime     string `json:"start_time"`      // "HH:MM"
	EndTime       string `json:"end_time"`         // "HH:MM"
	DurationMin   int    `json:"duration_min"`
	TravelFromPrev *TravelSegment `json:"travel_from_prev,omitempty"`
	Note          string `json:"note,omitempty"`
}

// TravelSegment represents travel between two activities.
type TravelSegment struct {
	Mode        TravelMode `json:"mode"`
	DurationMin int        `json:"duration_min"`
	DistanceM   int        `json:"distance_m"`
}

// TravelMode represents a mode of transportation.
type TravelMode string

const (
	TravelModeWalk    TravelMode = "walk"
	TravelModeTrain   TravelMode = "train"
	TravelModeBus     TravelMode = "bus"
	TravelModeTaxi    TravelMode = "taxi"
	TravelModeDriving TravelMode = "driving"
)

// NumDays returns the number of days in the itinerary.
func (it *Itinerary) NumDays() int {
	return int(it.EndDate.Sub(it.StartDate).Hours()/24) + 1
}
